File: packages/sdk/src/anchor/merkle.ts
import { createHash } from "node:crypto";

/**
 * SHA-256 Merkle tree for batching trace hashes.
 * Built in-memory — only the root goes onchain (via Anchor program PDA).
 * Individual proofs are stored in Convex for verification.
 */
export class MerkleTree {
  readonly leaves: Array<string>;
  private readonly layers: Array<Array<string>>;
  readonly root: string;

  constructor(hashes: Array<string>) {
    if (hashes.length === 0) {
      throw new Error("Cannot build Merkle tree from empty hash list");
    }

    // Copy and pad to power of 2 with zero-hashes
    this.leaves = [...hashes];
    const targetSize = nextPowerOf2(this.leaves.length);
    const zeroHash = "0".repeat(64);
    while (this.leaves.length < targetSize) {
      this.leaves.push(zeroHash);
    }

    this.layers = [this.leaves];
    let currentLayer = this.leaves;

    while (currentLayer.length > 1) {
      const nextLayer: Array<string> = [];
      for (let i = 0; i < currentLayer.length; i += 2) {
        nextLayer.push(hashPair(currentLayer[i]!, currentLayer[i + 1]!));
      }
      this.layers.push(nextLayer);
      currentLayer = nextLayer;
    }

    this.root = currentLayer[0]!;
  }

  /** Get the Merkle root as hex string */
  rootHex(): string {
    return this.root;
  }

  /** Get the Merkle root as Uint8Array (for onchain PDA) */
  rootBytes(): Uint8Array {
    return hexToBytes(this.root);
  }

  /**
   * Generate a proof for a leaf at the given index.
   * Returns sibling hashes needed to recompute root from leaf.
   */
  getProof(index: number): Array<string> {
    if (index < 0 || index >= this.leaves.length) {
      throw new Error(`Index ${index} out of bounds (${this.leaves.length} leaves)`);
    }

    const proof: Array<string> = [];
    let currentIndex = index;

    for (let layerIndex = 0; layerIndex < this.layers.length - 1; layerIndex++) {
      const layer = this.layers[layerIndex]!;
      const siblingIndex = currentIndex % 2 === 0 ? currentIndex + 1 : currentIndex - 1;

      if (siblingIndex < layer.length) {
        proof.push(layer[siblingIndex]!);
      }

      currentIndex = Math.floor(currentIndex / 2);
    }

    return proof;
  }

  /** Alias for getProof (already returns hex strings) */
  getProofHex(index: number): Array<string> {
    return this.getProof(index);
  }

  /**
   * Find the index of a leaf by its hex hash.
   * Returns -1 if not found.
   */
  indexOf(hashHex: string): number {
    return this.leaves.indexOf(hashHex);
  }
}

/**
 * Verify a Merkle proof: does this leaf + proof produce the expected root?
 * Standalone function — doesn't need the full tree.
 */
export function verifyMerkleProof(
  leafHex: string,
  proofHex: Array<string>,
  rootHex: string,
): boolean {
  let current = leafHex;

  for (const sibling of proofHex) {
    current = hashPair(current, sibling);
  }

  return current === rootHex;
}

function hashPair(a: string, b: string): string {
  // Consistent ordering: smaller hash on left
  const [left, right] = a <= b ? [a, b] : [b, a];
  const combined = Buffer.concat([hexToBytes(left!), hexToBytes(right!)]);
  return createHash("sha256").update(combined).digest("hex");
}

function hexToBytes(hex: string): Uint8Array {
  const bytes = new Uint8Array(hex.length / 2);
  for (let i = 0; i < hex.length; i += 2) {
    bytes[i / 2] = parseInt(hex.substring(i, i + 2), 16);
  }
  return bytes;
}

function nextPowerOf2(n: number): number {
  let power = 1;
  while (power < n) {
    power *= 2;
  }
  return power;
}

File: apps/server/internal/anchor/merkle.go
package anchor

import (
	"bytes"
	"crypto/sha256"
	"encoding/hex"
	"errors"
	"strconv"
)

// MerkleTree holds the tree layers built from leaf hashes.
// Only the root and proofs are needed after construction.
type MerkleTree struct {
	Leaves [][]byte
	Layers [][][]byte
	Root   []byte
}

// BuildMerkleTree constructs a SHA-256 Merkle tree from hex-encoded trace hashes.
// Pads to next power of 2 with zero-hashes.
func BuildMerkleTree(hexHashes []string) (*MerkleTree, error) {
	if len(hexHashes) == 0 {
		return nil, errors.New("cannot build Merkle tree from empty hash list")
	}

	leaves := make([][]byte, len(hexHashes))
	for i, h := range hexHashes {
		b, err := hex.DecodeString(h)
		if err != nil {
			return nil, errors.New("invalid hex hash at index " + strconv.Itoa(i))
		}
		leaves[i] = b
	}

	// Pad to power of 2
	targetSize := nextPowerOf2(len(leaves))
	zeroHash := make([]byte, 32)
	for len(leaves) < targetSize {
		leaves = append(leaves, zeroHash)
	}

	layers := [][][]byte{leaves}
	currentLayer := leaves

	for len(currentLayer) > 1 {
		var nextLayer [][]byte
		for i := 0; i < len(currentLayer); i += 2 {
			nextLayer = append(nextLayer, hashPair(currentLayer[i], currentLayer[i+1]))
		}
		layers = append(layers, nextLayer)
		currentLayer = nextLayer
	}

	return &MerkleTree{
		Leaves: leaves,
		Layers: layers,
		Root:   currentLayer[0],
	}, nil
}

// RootHex returns the root hash as hex string.
func (t *MerkleTree) RootHex() string {
	return hex.EncodeToString(t.Root)
}

// GetProof returns the Merkle proof (sibling hashes) for a leaf at the given index.
func (t *MerkleTree) GetProof(index int) ([][]byte, error) {
	if index < 0 || index >= len(t.Leaves) {
		return nil, errors.New("index out of bounds")
	}

	var proof [][]byte
	currentIndex := index

	for layerIdx := 0; layerIdx < len(t.Layers)-1; layerIdx++ {
		layer := t.Layers[layerIdx]
		siblingIndex := currentIndex ^ 1 // flip last bit

		if siblingIndex < len(layer) {
			proof = append(proof, layer[siblingIndex])
		}

		currentIndex /= 2
	}

	return proof, nil
}

// GetProofHex returns the proof as hex strings.
func (t *MerkleTree) GetProofHex(index int) ([]string, error) {
	proof, err := t.GetProof(index)
	if err != nil {
		return nil, err
	}

	hexProof := make([]string, len(proof))
	for i, p := range proof {
		hexProof[i] = hex.EncodeToString(p)
	}
	return hexProof, nil
}

// VerifyProof checks that a leaf + proof produces the expected root.
func VerifyProof(leafHex string, proofHex []string, rootHex string) bool {
	current, err := hex.DecodeString(leafHex)
	if err != nil {
		return false
	}

	for _, siblingHex := range proofHex {
		sibling, err := hex.DecodeString(siblingHex)
		if err != nil {
			return false
		}
		current = hashPair(current, sibling)
	}

	return hex.EncodeToString(current) == rootHex
}

func hashPair(a, b []byte) []byte {
	// Consistent ordering: smaller hash on left
	if bytes.Compare(a, b) > 0 {
		a, b = b, a
	}
	h := sha256.New()
	h.Write(a)
	h.Write(b)
	return h.Sum(nil)
}

func nextPowerOf2(n int) int {
	power := 1
	for power < n {
		power *= 2
	}
	return power
}
